crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,150 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
         })
         .collect()
 }
+
+// Semaphore like implementation that allows both shrinking and expanding
+// the max permits.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphore(Arc<SemShared>);
+
+// The permit.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphorePermit(Arc<SemShared>);
+
+// The state shared between the semaphore and the outstanding permits.
+#[derive(Debug)]
+struct SemShared {
+    // The tuple holds (current usage, current max capacity)
+    current: Mutex<SemState>,
+
+    // To signal waiters for permits to be available
+    notify: Notify,
+}
+
+// Current state.
+#[derive(Debug)]
+struct SemState {
+    // The current capacity
+    capacity: usize,
+
+    // The current outstanding permits
+    usage: usize,
+}
+
+impl SemShared {
+    fn new(capacity: usize) -> Self {
+        Self {
+            current: Mutex::new(SemState { capacity, usage: 0 }),
+            notify: Notify::new(),
+        }
+    }
+
+    // Allocates a permit if available.
+    // Returns true if a permit was allocated, false otherwise
+    async fn alloc_one(&self) -> bool {
+        let mut current = self.current.lock().await;
+        if current.usage < current.capacity {
+            current.usage += 1;
+            true
+        } else {
+            false
+        }
+    }
+
+    // Returns a permit back to the free pool, notifies waiters if needed.
+    async fn free_one(&self) {
+        let should_notify = {
+            let mut current = self.current.lock().await;
+            if current.usage > 0 {
+                current.usage -= 1;
+            } else {
+                panic!(
+                    "SemShared::free_one(): invalid free, current = {:?}",
+                    current
+                );
+            }
+
+            // Notify only if usage fell below the current capacity.
+            // For example: if the previous capacity was 100, and current capacity
+            // is 50, this will wait for usage to fall below 50 before any waiters
+            // are notified.
+            current.usage < current.capacity
+        };
+        if should_notify {
+            self.notify.notify_waiters();
+        }
+    }
+
+    // Expands the max capacity by delta, and notifies any waiters of the newly available
+    // free permits.
+    async fn expand(&self, delta: usize) {
+        {
+            let mut current = self.current.lock().await;
+            current.capacity += delta;
+        }
+        self.notify.notify_waiters();
+    }
+
+    // Shrinks the max capacity by delta
+    async fn shrink(&self, delta: usize) -> Result<(), String> {
+        let mut current = self.current.lock().await;
+        if current.capacity > delta {
+            current.capacity -= delta;
```

Comment:
There is `checked_sub` method just for this case. Just a minor nitpick about idiomatic Rust usage.